### PR TITLE
FIX: Locations max issue

### PR DESCRIPTION
**Pull request by alexandre-merle**
For parts with many locations, setPartRanges was going one location more
than needed, resulting in sending more data than the content-length.

### CODE REVIEW

**Comment by alexandre-merle** [comment]
@ironman-machine try

**Comment by alexandre-merle** [comment]
failed, but fixing the first issue

**Comment by alexandre-merle** [comment]
@ironman-machine
SCALITY_BACKBEAT_BRANCH=test/aws-sdk-version
try

**Comment by alexandre-merle** [comment]
@ironman-machine try

**Comment by alexandre-merle** [comment]
@ironman-machine 
SCALITY_INTEGRATION_BRANCH=test/backbeat-issue
try

**Comment by alexandre-merle** [comment]
@ironman-machine try

**Comment by alexandre-merle** [comment]
@ironman-machine try

**Comment by alexandre-merle** [comment]
@ironman-machine r+

**Comment by alexandre-merle** [comment]
oups, miss command

**Comment by alexandre-merle** [comment]
@ironman-machine try

**Comment by alexandre-merle** [comment]
@ironman-machine 
SCALITY_INTEGRATION_BRANCH=test/backbeat-wait-complete
try

**Comment by LaurenSpiegel** [comment]
How did this come up as an issue?

File: lib/api/apiUtils/object/setPartRanges.js [hunk 1]
```diff
@@ -37,7 +37,7 @@ function setPartRanges(dataLocations, outerRange) {
     const max = end - begin + 1;
     let total = 0;
     for (let i = 0; i < dataLocations.length; i++) {
-        if (total >= max) {
+        if (total >= (max - 1)) {
```
Comment (LaurenSpiegel) [inline lib/api/apiUtils/object/setPartRanges.js]: On line 36, it says we are adding 1 because range is inclusive. Why this change?

**Comment by alexandre-merle** [comment]
@LaurenSpiegel i have done a lot of tests this week end, and this is an issue i see in the container 5 of backbeat, this PR fix the issue.
For the context, backbeat was requesting a location start: 11, size: 11 and get back 12 character from S3, which leads to instable connections, 'data backend' econnreset, and other stuff

**Comment by alexandre-merle** [comment]
@LaurenSpiegel the related integration tests you can look into is http://ci.ironmann.io/gh/scality/Integration/16518#artifacts, you have a network trace inside it showing the bug

File: lib/api/apiUtils/object/setPartRanges.js (hunk 1 above)
Comment (jonathan-gramain) [inline lib/api/apiUtils/object/setPartRanges.js]: This seems wrong, the problem is that we don't specify clearly the convention of whether `end` is inclusive or exclusive, ending up having different callers using different conventions. See e.g. https://github.com/scality/S3/blob/master/lib/api/objectGet.js#L188, where the part end boundary is exclusive but `byteRange` is inclusive. I haven't gone through other callers, my gut feeling is that it's meant to be inclusive (hence the comment about `+1` in `setPartRanges`).

I suggest we should go through those callers and fix if they don't follow the convention (hopefully only the part get is in this case but we never know), and document whether it's inclusive or exclusive in JSDoc of `setPartRanges` to avoid future mistakes.

Comment (jonathan-gramain) [reply lib/api/apiUtils/object/setPartRanges.js]: Btw, thanks @alexandre-merle for nailing down the issue :+1:

**Comment by jonathan-gramain** [comment]
Please review https://github.com/scality/S3/pull/974, if that's ok and passes the tests it should replace this PR.
